pkg/apis/common/v1/association.go

```diff
@@ -48,23 +98,33 @@ type Association interface {
 
 	// AssociatedType returns a string describing the type of the target resource (elasticsearch most of the time)
 	// It is mostly used to build some other strings depending on the type of the targeted resource.
-	AssociatedType() string
+	AssociatedType() AssociationType
 
 	// Reference to the associated resource. If defined with a Name then the Namespace is expected to be set in the returned object.
 	AssociationRef() ObjectSelector
 
-	// AssociationConfAnnotationName is the name of the annotation used to define the config for the associated resource.
+	// AssociationConfAnnotationNameBase is the name of the annotation used to define the config for the associated resource.
 	// It is used by the association controller to store the configuration and by the controller which is
 	// managing the associated resource to build the appropriate configuration.
-	AssociationConfAnnotationName() string
+	AssociationConfAnnotationNameBase() string
 
 	// Configuration
 	AssociationConf() *AssociationConf
 	SetAssociationConf(*AssociationConf)
 
-	// Status
-	AssociationStatus() AssociationStatus
-	SetAssociationStatus(status AssociationStatus)
+	// Id allows to distinguish between many associations of the same type
+	ID() int
```

Comment:
I feel like an ID should uniquely identify an association. That would also save us a lot of key construction later on in the code.

Reply:
I wonder if this could even be something like

```golang
type AssocationID struct {
	Type       AssociationType
	Associated ObjectSelector
	SeqNo      int
}

func (aid AssocationID) AnnotationName() string {
	typ := "unknown"
	switch aid.Type {
	case ElasticsearchAssociationType:
		typ = ElasticsearchConfigAnnotationNameBase
	case KibanaAssociationType:
		typ = KibanaAssociationType
	}
	return fmt.Sprintf("%s-%d", typ, aid.SeqNo)
}

func (aid AssocationID) String() string {
	return fmt.Sprintf("%s.%s.%s.%d", aid.Type, aid.Associated.Namespace, aid.Associated.Name, aid.SeqNo)
}
```

Reply:
See my [other](https://github.com/elastic/cloud-on-k8s/pull/4036#discussion_r543458804) comment.

Reply:
Moved annotation name generation to `Association` struct.